Support multiple local directories

somebody1234 opened the pull request:
### Pull Request Description

- Add support for multiple local directories
- Save primary local root directory in `localStorage`
  - Fix primary local root directory not updating in some conditions
- Save additional local root directories in `localStorage`
- Add buttons to browse for and reset primary local root directory

### Important Notes
None

### Screenshots

Changes to sidebar:
![image](https://github.com/user-attachments/assets/bfe202fc-e0d6-4c0c-894a-8d21786ed97e)

Changes to "local" settings page:
![image](https://github.com/user-attachments/assets/f1db5b91-1812-4573-a90a-26e3b6302390)

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [x] ~~Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.~~
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] ~~Unit tests have been written where possible.~~

somebody1234 committed:
User/team spaces

somebody1234 committed:
Explicit root directory id

somebody1234 committed:
Move Category Switcher logic to CategorySwitcherItem

somebody1234 committed:
Move category info from objects to props

somebody1234 committed:
WIP: Initial UI support for displaying users and teams in sidebar

Support for actually switching to the directories will come later.

somebody1234 committed:
WIP: Refactor `Category` into a struct

somebody1234 committed:
Fix type error

somebody1234 committed:
Merge branch 'develop' into wip/sb/user-team-spaces

somebody1234 committed:
Fix final type errors

somebody1234 committed:
Fix root path of user/team directories

somebody1234 committed:
Fix category switcher

somebody1234 committed:
Fix/suppress lint errors

somebody1234 committed:
Add "My Files" category

somebody1234 committed:
Merge branch 'develop' into wip/sb/user-team-spaces

somebody1234 committed:
Avoid using empty string as root directory id

somebody1234 committed:
Merge branch 'develop' into wip/sb/user-team-spaces

somebody1234 committed:
Merge branch 'develop' into wip/sb/user-team-spaces

somebody1234 committed:
Fix root directory id for user and team folders

somebody1234 committed:
Fix typo

PabloBuchu committed:
update users/teams directories ids

PabloBuchu committed:
Revert "update users/teams directories ids"

This reverts commit 8050dabfabb88472e5174431ad778733beb3b2e7.

somebody1234 committed:
Merge branch 'develop' into wip/sb/user-team-spaces

PabloBuchu committed:
change hardoced users and teams ids

somebody1234 committed:
Merge branch 'develop' into wip/sb/user-team-spaces

somebody1234 committed:
Change "owner" permission based on owner of category

somebody1234 committed:
Hide "add assets" context menu on read-only directories

somebody1234 committed:
Update "self" permission extraction

somebody1234 committed:
Prevent opening assets context menu in organization root directory

somebody1234 committed:
Disallow creating assets in org root directory via Drive Bar

somebody1234 committed:
Switch back to listing `Users` and `Teams` directories to enumerate folders because the user does not have access to every user/team

somebody1234 committed:
Minor changes to text

somebody1234 committed:
Fix lint errors

somebody1234 committed:
Minimal fixes for mock APIs

somebody1234 committed:
Fix E2E tests

somebody1234 committed:
Merge branch 'develop' into wip/sb/user-team-spaces

somebody1234 committed:
Merge branch 'develop' into wip/sb/user-team-spaces

somebody1234 committed:
Disable Drive Bar buttons instead of replacing with a button to go to My Files

somebody1234 committed:
Do not delete `/Users` and `/Teams` when clearing trash

somebody1234 committed:
Add visual tooltip for disabled "create asset" buttons

somebody1234 committed:
Enable asset actions again when a directory is selected

somebody1234 committed:
Disable creating assets in `/Users` and `/Teams`

somebody1234 committed:
Use permissions to determine whether to disable asset creation

somebody1234 committed:
Disable asset creation based on root directory only when no directory is selected

somebody1234 committed:
Merge branch 'develop' into wip/sb/user-team-spaces

somebody1234 committed:
Edit prettier config (#10648)

somebody1234 committed:
Edit prettier config (#10648)

somebody1234 committed:
Reenable "Local" settings tab

somebody1234 committed:
Merge branch 'develop' into wip/sb/user-team-spaces

somebody1234 committed:
Merge branch 'wip/sb/reenable-local-settings' into wip/sb/multiple-local-directories

somebody1234 committed:
Initial support for custom local directories

somebody1234 committed:
Add button to add new local directory

somebody1234 committed:
Fix listing extra local directories

somebody1234 committed:
Save local root directory to localStorage

somebody1234 committed:
Sync and reset Local root directory updates

somebody1234 committed:
Properly sync local root directory with Data Catalog

somebody1234 committed:
Remove unnecessary mutation around `updateLocalRootDirectory`

somebody1234 committed:
Rename `useLocalStorageKey` to `useLocalStorageState`

somebody1234 committed:
Add button to browse for new root folder

somebody1234 committed:
Merge branch 'develop' into wip/sb/multiple-local-directories

somebody1234 committed:
Replace owner when moving between users/groups

somebody1234 committed:
Replace owner when copying between users/groups

somebody1234 committed:
Disable pasting from Team to User via context menu

somebody1234 committed:
Disable dragging assets from teams to users

somebody1234 committed:
Fix self permission detection for user groups

somebody1234 committed:
Format TS code

somebody1234 committed:
Merge branch 'develop' into wip/sb/user-team-spaces

somebody1234 committed:
Format TS code

somebody1234 committed:
Merge branch 'wip/sb/user-team-spaces' into wip/sb/multiple-local-directories

somebody1234 committed:
Fix invariant error

MrFlashAccount commented on app/dashboard/src/components/AriaComponents/Button/ButtonGroup.tsx:
```diff
@@ -51,11 +51,11 @@ const STYLES = twv.tv({
 
 /** A group of buttons. */
-export function ButtonGroup(props: ButtonGroupProps) {
+function ButtonGroupInternal(props: ButtonGroupProps, ref: React.ForwardedRef<HTMLDivElement>) {
```
Can we use the `export const ButtonGroup = React.forwardRef(function ButtonGroup(props)` format instead of this one? With this approach you'll see better display names in the devtools

MrFlashAccount commented on app/dashboard/src/App.tsx:
```diff
@@ -278,6 +282,24 @@ function AppRouter(props: AppRouterProps) {
     [projectManagerInstance],
   )
 
+  React.useEffect(() => {
```
Prefer please `useSyncExternalStore` : https://react.dev/reference/react/useSyncExternalStore#reference

somebody1234 replied:
i don't know if it's appropriate here, `localBackend.rootPath` is not reactive

somebody1234 replied:
(note that we already use `useSyncExternalStore` elsewhere where we *do* need the reactive updates.)

MrFlashAccount replied:
I see we subscribe in the useEffect, it'd be better to use useSyncExternalStore for that case

somebody1234 replied:
sure, but then the parent component will re-render, no? in which case we'd have to move this `useEffect` to its own child component just to avoid that. not that i'm against that, i just figured it wasn't good practice to create a child component just so we can use `useSyncExternalStore`

it's also worth noting that we have similar subscribe functions in a useEffect for our shortcut handling (and i do believe it is standard practice to subscribe in effects anyway...)

and again i need to note that this is not reactive state we are changing

MrFlashAccount replied:
> and I do believe it is standard practice to subscribe in effects anyway...

Not exactly. `useSyncExternalStore` subscribes during the render, `useEffect` - after the render, so useSyncExternalStore will always provide fresh data. useSyncExternalStore supports selection: https://thisweekinreact.com/articles/useSyncExternalStore-the-underrated-react-api, and it's a dedicated API for these situations. 

> It's also worth noting that we have similar subscribe functions in a useEffect

If we're not reading in subscribe (e.g. assigning event listeners), then it's okay to use `useEffect`

In that case, we might get an out-of-sync state when we want to retrieve projects but the path is not updated yet. We should be careful with that. It makes it hard to catch async errors.

In general, I don't like this solution. Why we're not changing the path directly in a callback or during a render?
And I'm double down on moving this logic away from the `App` component. Ideally, App should only create deps and the root tree.

somebody1234 replied:
> Why we're not changing the path directly in a callback or during a render?

because the path should stay in sync with the value in `localStorage`. idk, we could maybe move this event listener into the backend provider, but i'm not sure that's a complete solution

i guess (theoretically at least) we aren't actually using the root directory anymore because we always provide a directory name to all API calls so we could probably remove this code without repercussions.

but anyway... thoughts? is `useSyncExternalStore` in a nested component the solution here then?

MrFlashAccount replied:
If we don't need the root path, let's remove it.

somebody1234 replied:
see current version of code: <https://github.com/enso-org/enso/pull/10601/files?w=1#diff-1496d10f19e32b61e5063c49393ff603adf03db009ffcfae3c093cb58f7cdc21>
lmk if i should rename the file to something else (e.g. revert but add docs)/improve docs/something else

somebody1234 replied:
oops replied in wrong thread

somebody1234 replied:
unfortunately it seems like we do still need to sync `rootPath` for the time being because we are currently manually calculating paths - i think this is blocked until cloud backend starts giving us paths? see 4f31636 though for the changes

MrFlashAccount replied:
LGTM

MrFlashAccount commented on app/dashboard/src/components/dashboard/column/SharedWithColumn.tsx:
```diff
@@ -45,21 +45,16 @@ export default function SharedWithColumn(props: SharedWithColumnPropsInternal) {
-  const { user } = authProvider.useNonPartialUserSession()
+  const { user } = authProvider.useFullUserSession()
   const dispatchAssetEvent = eventListProvider.useDispatchAssetEvent()
-
```
Can we please keep semantic delimiters in place?

Having blank lines improve readability and group similar things together

somebody1234 replied:
sure, but how do we determine when we need a delimiter? do we need one between `isFeatureUnderPaywall` and its usage (as was previously the case)?

MrFlashAccount replied:
> sure, but how do we determine when we need a delimiter? do we need one between isFeatureUnderPaywall and its usage (as was previously the case)?

Subjectively. Listen to your heart 😊

MrFlashAccount commented on app/dashboard/src/globals.d.ts:
```diff
@@ -111,6 +111,30 @@ interface ProjectManagementApi {
+// === File Browser API ===
+// ========================
+
+/** `window.fileBrowserApi` is a context bridge to the main process, when we're running in an
```
Hope that we'll eventually collect all these API under a single field to avoid potential mismatches with the new API/extensions/even ids(https://stackoverflow.com/questions/3434278/do-dom-tree-elements-with-ids-become-global-properties)

MrFlashAccount replied:
Ideally it'd be great to use `Symbol` for that

somebody1234 replied:
> potential mismatches

not possible, all our apis are suffixed with `Api`. although it is a somewhat legitimate concern it's unlikely to happen in practice

> Ideally it'd be great to use `Symbol` for that

i guess you mean `Symbol.for`... which does make sense i guess? although the major downside is that there's no way to represent a `Symbol.for` in a type-safe manner (and i think it's a little overkill. we can just prefix with `enso` if we're worried about conflicts)

also note that this API already exists and was in fact introduced by GUI. although it's true that most of the APIs do exist for dashboard...

MrFlashAccount replied:
> not possible, all our APIs are suffixed with Api. although it is a somewhat legitimate concern it's unlikely to happen in practice

Oh, how many times I've heard this 😭

It'd be great to keep all electron-related APIs under a single object. At least it'll be much easier to track that API

somebody1234 replied:
i mean, personally i've never ever seen a spec that defined a global containing `Api` in its name, but i could be entirely wrong.

obviously it's preferable for everything to be under a single variable, i'm just pointing out that realistically i don't think it will be an issue for the next couple years at least.

MrFlashAccount replied:
Yeah, currently it's not an issue and we can live with that, I just wish things were different 😊

somebody1234 replied:
yeah definitely. i think this kind of thing is out of scope of this PR, i guess things like this should be opened as issues?

somebody1234 replied:
- opened as https://github.com/enso-org/enso/issues/10699

MrFlashAccount commented on app/dashboard/src/layouts/CategorySwitcher/Category.ts:
```diff
@@ -1,34 +1,91 @@
 
 /** The categories available in the category switcher. */
-enum Category {
+export enum CategoryType {
```
Maybe it's time to split local categories with the cloud ones?

somebody1234 replied:
we *could*, but is there an advantage, when we want to handle both local and cloud categories in every component?

MrFlashAccount replied:
If we can't merge 2 enums then we shouldn't use enums.

somebody1234 replied:
again my point is that there is no advantage in splitting this enum in the first place. i have nothing against splitting this, but there is no advantage that i can see, it's not like it lets us simplify any code

MrFlashAccount replied:
I see these advantages:

1. Easier to understand if a category is local or cloud
2. Less error prone when it comes to the functions like `isCloud`, is `isLocal`
3. Easier to extend in the future

somebody1234 replied:
i mean, i switched and now all the category types are inside their respective objects (instead of as a string union). not sure if it's ideal but `isCloud` and `isLocal` are able to use the new object unions, but idk if 1 and 3 have been addressed

(note that i'm not against adding a string union, i'm just not sure how to make sure they're in sync with the object types. but also they don't seem necessary yet so maybe we can leave it for later?)

MrFlashAccount commented on app/dashboard/src/layouts/Settings/settingsData.tsx:
```diff
@@ -238,9 +239,43 @@ export const SETTINGS_TAB_DATA: Readonly<Record<SettingsTabType, SettingsTabData
-            setValue: (context, value) => context.updateLocalRootPath(value),
+            setValue: async (context, value) => {
+              context.updateLocalRootPath(value)
+              await Promise.resolve()
```
btw you can just mark the function as async, no need to `await Promise.resolve()`

somebody1234 replied:
we can't because of eslint (while technically we could remove the rule, i think it makes sense to leave it out since it's helpful in the vast majority of cases

MrFlashAccount commented on app/dashboard/src/pages/dashboard/Dashboard.tsx:
```diff
@@ -120,16 +121,59 @@ function DashboardInner(props: DashboardProps) {
     (value): value is Category => {
-      if (array.includes(Object.values(Category), value)) {
-        return categoryModule.isLocal(value) ? localBackend != null : true
+      if (typeof value !== 'object' || value == null) {
```
can we use Zod schema with additional `refine` here? Looks overcomplicated

MrFlashAccount commented on app/dashboard/src/providers/DriveProvider.tsx:
```diff
@@ -0,0 +1,84 @@
+import invariant from 'tiny-invariant'
+import * as zustand from 'zustand'
+
+import { type DirectoryAsset } from 'enso-common/src/services/Backend'
```
We have backend re-export in the project, shall we use the reexport or import directly from the `enso-common`?

somebody1234 replied:
dunno tbh. i guess importing directly makes more sense, that way we don't have duplicate imports (because the `enso-common` exports will show up in autoocomplete no matter what). thoughts?

(in which case i guess ideally we want to remove re-exports)

MrFlashAccount replied:
If we're not going to have wrappers around common backend, then let's use direct imports

somebody1234 committed:
Address CR

somebody1234 committed:
Use zod schema for category

somebody1234 committed:
Remove enum from Category

somebody1234 committed:
Switch from `Effect` to `SyncExternalStore` to sync local backend root path

somebody1234 committed:
Fix `isCloudCategory` and `isLocalCategory`

somebody1234 committed:
Merge branch 'develop' into wip/sb/user-team-spaces

somebody1234 committed:
Merge branch 'wip/sb/user-team-spaces' into wip/sb/multiple-local-directories

somebody1234 committed:
Merge branch 'develop' into wip/sb/user-team-spaces

somebody1234 committed:
Fix lint errors

somebody1234 committed:
Merge branch 'wip/sb/user-team-spaces' into wip/sb/multiple-local-directories

somebody1234 committed:
Remove `triggerRerender` from LocalStorage

PabloBuchu committed:
Merge branch 'develop' into wip/sb/user-team-spaces

somebody1234 committed:
Try to fix automatically generated permission

PabloBuchu committed:
Merge branch 'wip/sb/user-team-spaces' of https://github.com/enso-org/enso into wip/sb/user-team-spaces

somebody1234 committed:
Properly update selected target directory again

PabloBuchu committed:
Merge branch 'wip/sb/user-team-spaces' of https://github.com/enso-org/enso into wip/sb/user-team-spaces

somebody1234 committed:
Merge branch 'develop' into wip/sb/user-team-spaces

somebody1234 committed:
Merge branch 'wip/sb/user-team-spaces' into wip/sb/multiple-local-directories

somebody1234 committed:
Merge branch 'develop' into wip/sb/user-team-spaces

somebody1234 committed:
Remove `components/Input`

somebody1234 committed:
Fix E2E tests

somebody1234 committed:
Fix automatically detected initial permissions

somebody1234 committed:
Merge branch 'wip/sb/user-team-spaces' into wip/sb/multiple-local-directories

somebody1234 committed:
Merge branch 'develop' into wip/sb/multiple-local-directories

somebody1234 committed:
Update rust-toolchain hash in Nix to unbreak wasm builds when using Nix

PabloBuchu commented:
1. Start local project (you will be redirected to graph view). Go back to dashboard -> you wont find newly created project on the list (it will show up after reload)
2. Something is broken when trying to create project in subdirectory. I am selecting a folder but `New empty project` is disabled
3. I am able to drag n drop things between root folders although they are disappearing from source but not landing in destination (which is easier fixing moving or disable drag n drop?)
4. I can't make it to go away...
<img width="409" alt="Screenshot 2024-08-22 at 13 30 31" src="https://github.com/user-attachments/assets/504a15fb-8662-4a0f-9447-1d193860f19d">
5. Reload on opened project results in `error try again`

https://github.com/user-attachments/assets/504f56d6-17dd-4c94-8210-dc7783a8043a

somebody1234 committed:
Adjust UI shown when failing to connect to PM

somebody1234 committed:
Properly align settings icon to right side for Local category

somebody1234 committed:
Disable permissions check for adding assets for local categories

somebody1234 commented:
1. seems to be fixed locally. can confirm that this is caused by logic that is no longer present in #10775
2. yeah i need to disable the permissions check on the local backend, fixed locally
3. fixed locally (?!) - the logic was previously completely absent, just added it and by some miracle it seems to Just Work
  - logic for determining whether a category was a valid drop target was previously broken, fixed as part of this
5. could repro before but can no longer repro?
6. seperate issue, will address in another PR

- i suspect 1. and 3. may be partially addressed (or at least made easier) by #10775


1. is really weird because it for sure shows up *before* switching to the editor tab...

somebody1234 committed:
Prevent querying user and team space folders when user does not have them

somebody1234 committed:
Merge branch 'develop' into wip/sb/multiple-local-directories

somebody1234 committed:
Add label to "remove" button for local directories

somebody1234 committed:
Fix newly opened project disappearing from Local Backend

somebody1234 committed:
Switch from delete icon to minus icon for removing local favorite
folders

somebody1234 committed:
Fix logic determining whether a category is a valid drop target

somebody1234 committed:
Fix moving between local directories

somebody1234 committed:
Fix errors

somebody1234 committed:
Assets can be dragged out of recent category

somebody1234 commented:
@PabloBuchu should be ready for re-QA. the changes should also *theoretically* fix dragging things between user/team/cloud directories but not 100% sure it works

PabloBuchu commented:
Would be nice to limit chars to fit the column.. other that that looks good and can be merged

<img width="396" alt="Screenshot 2024-08-28 at 10 57 36" src="https://github.com/user-attachments/assets/0f2faf96-e983-4a3c-80f9-580f373eb51c">

somebody1234 committed:
WIP

somebody1234 committed:
Truncate category labels
